File: tmail-backend/apps/memory/src/main/scala/com/linagora/tmail/james/app/modules/jmap/MemoryEmailAddressContactModule.scala
package com.linagora.tmail.james.app.modules.jmap

import com.google.inject.multibindings.ProvidesIntoSet
import com.google.inject.name.Named
import com.google.inject.{AbstractModule, Provides, Singleton}
import com.linagora.tmail.james.jmap.EmailAddressContactInjectKeys
import com.linagora.tmail.james.jmap.contact.{EmailAddressContactListener, InMemoryEmailAddressContactSearchEngineModule}
import org.apache.james.events.EventBus
import org.apache.james.lifecycle.api.Startable
import org.apache.james.utils.{InitializationOperation, InitilizationOperationBuilder}

class MemoryEmailAddressContactModule extends AbstractModule {

  override def configure(): Unit = {
    install(new InMemoryEmailAddressContactSearchEngineModule)
  }

  @ProvidesIntoSet
  def registerListener(@Named(EmailAddressContactInjectKeys.AUTOCOMPLETE) eventBus: EventBus,
                       emailAddressContactListener: EmailAddressContactListener): InitializationOperation = {
    InitilizationOperationBuilder.forClass(classOf[EmailAddressContactEventLoader])
      .init(() => eventBus.register(emailAddressContactListener))
  }

  @Provides
  @Singleton
  @Named(EmailAddressContactInjectKeys.AUTOCOMPLETE)
  def provideInVMEventBus(eventBus: EventBus): EventBus = eventBus

}

class EmailAddressContactEventLoader extends Startable

File: tmail-backend/apps/memory/src/main/java/com/linagora/tmail/james/app/modules/jmap/MemoryEmailAddressContactModule.java
package com.linagora.tmail.james.app.modules.jmap;

import org.apache.james.events.EventBus;
import org.apache.james.lifecycle.api.Startable;
import org.apache.james.utils.InitializationOperation;
import org.apache.james.utils.InitilizationOperationBuilder;

import com.google.inject.AbstractModule;
import com.google.inject.Provides;
import com.google.inject.Singleton;
import com.google.inject.multibindings.ProvidesIntoSet;
import com.google.inject.name.Named;
import com.linagora.tmail.james.jmap.EmailAddressContactInjectKeys;
import com.linagora.tmail.james.jmap.contact.EmailAddressContactListener;
import com.linagora.tmail.james.jmap.contact.InMemoryEmailAddressContactSearchEngineModule;

public class MemoryEmailAddressContactModule extends AbstractModule {

    @Override
    protected void configure() {
        install(new InMemoryEmailAddressContactSearchEngineModule());
    }

    @ProvidesIntoSet
    public InitializationOperation registerListener(
            @Named(EmailAddressContactInjectKeys.AUTOCOMPLETE) EventBus eventBus,
            EmailAddressContactListener emailAddressContactListener) {
        return InitilizationOperationBuilder
                .forClass(EmailAddressContactEventLoader.class)
                .init(() -> eventBus.register(emailAddressContactListener));
    }

    @Provides
    @Singleton
    @Named(EmailAddressContactInjectKeys.AUTOCOMPLETE)
    public EventBus provideInVMEventBus(EventBus eventBus) {
        return eventBus;
    }

    public static class EmailAddressContactEventLoader implements Startable {

    }
}


